core/server/common/src/main/java/alluxio/master/journal/JournalTool.java

```diff
@@ -239,4 +194,299 @@ private static void usage() {
         "Read an Alluxio journal and write it to a directory in a human-readable format.", OPTIONS,
         "", true);
   }
+
+  /**
+   * An abstract class for journal dumpers.
+   */
+  abstract static class AbstractJournalDumper {
+    protected final String mMaster;
+    protected final long mStart;
+    protected final long mEnd;
+    protected final String mInputDir;
+    protected final String mOutputDir;
+    protected final String mCheckpointsDir;
+    protected final String mJournalEntryFile;
+
+    /**
+     * @param master journal master
+     * @param start journal start sequence
+     * @param end journal end sequence
+     * @param outputDir output dir for journal dump
+     * @param inputDir input dir for journal files
+     */
+    public AbstractJournalDumper(String master, long start, long end, String outputDir,
+        String inputDir) throws IOException {
+      mMaster = master;
+      mStart = start;
+      mEnd = end;
+      mInputDir = inputDir;
+      mOutputDir = outputDir;
+      mCheckpointsDir = PathUtils.concatPath(outputDir, "checkpoints");
+      mJournalEntryFile = PathUtils.concatPath(outputDir, "edits.txt");
+
+      // Ensure output directory structure.
+      Files.createDirectories(Paths.get(mOutputDir));
+      Files.createDirectories(Paths.get(mCheckpointsDir));
+    }
+
+    /**
+     * Dumps journal.
+     */
+    abstract void dumpJournal() throws Throwable;
+
+    /**
+     * Used to read checkpoint streams.
+     *
+     * @param checkpoint the checkpoint stream
+     * @param path persistence path
+     * @throws IOException
+     */
+    protected void readCheckpoint(CheckpointInputStream checkpoint, Path path) throws IOException {
+      LOG.debug("Reading checkpoint of type %s to %s%n", checkpoint.getType().name(), path);
+      switch (checkpoint.getType()) {
+        case COMPOUND:
+          readCompoundCheckpoint(checkpoint, path);
+          break;
+        case ROCKS:
+          readRocksCheckpoint(checkpoint, path);
+          break;
+        default:
+          readRegularCheckpoint(checkpoint, path);
+          break;
+      }
+    }
+
+    private void readCompoundCheckpoint(CheckpointInputStream checkpoint, Path path)
+        throws IOException {
+      Files.createDirectories(path);
+      CompoundCheckpointReader reader = new CompoundCheckpointReader(checkpoint);
+      Optional<Entry> entryOpt;
+      while ((entryOpt = reader.nextCheckpoint()).isPresent()) {
+        Entry entry = entryOpt.get();
+        Path checkpointPath = path.resolve(entry.getName().toString());
+        LOG.debug("Reading checkpoint for %s to %s%n", entry.getName(), checkpointPath);
+        readCheckpoint(entry.getStream(), checkpointPath);
+      }
+    }
+
+    private void readRocksCheckpoint(CheckpointInputStream checkpoint, Path path)
+        throws IOException {
+      TarballCheckpointReader reader = new TarballCheckpointReader(checkpoint);
+      reader.unpackToDirectory(path);
```

Comment:
Does this output human readable data?

Reply:
No. We might need another tool for dumping RocksDB content.

Reply:
okay, this won't be addressed in this PR?

Reply:
No. That's out of scope for this PR.